docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md

```diff
@@ -0,0 +1,101 @@
+        worksheet.write(row_idx, 7, record.row_number)
+        worksheet.write(row_idx, 8, str(ParserErrorCategoryChoices(record.error_type).label))
+```
+
+The three helper functions: `format_error_msg`, `friendly_names`, `internal_names` used to write the appropriate fields can be seen below.
+
+```python
+def format_error_msg(error_msg, fields_json):
+    """Format error message."""
+    for key, value in fields_json['friendly_name'].items():
+        error_msg = error_msg.replace(key, value) if value else error_msg
+    return error_msg
+
+
+def friendly_names(fields_json):
+    """Return comma separated string of friendly names."""
+    return ','.join([i for i in fields_json['friendly_name'].values()])
+
+
+def internal_names(fields_json):
+    """Return comma separated string of internal names."""
+    return ','.join([i for i in fields_json['friendly_name'].keys()])
+```
+
+### Prioritized Errors
+[This OneNote](https://gorafttech.sharepoint.com/:o:/s/TDRSResearchDesign/EnIa1Mn4v7pOskW7BLomXhIBxUMlYLRU_f1C0dxemW7dWw?e=m0rNyI) is invaluable to the implementation of prioritized errors. Prioritizing errors could be a very large and technically challenging feature involving new migrations, validation/validator refactors, etc... However, this can all be avoided by making a key insight for each of the category two and category three validators by way of OFA's requirements for them. For the category two case, the OneNote document generically specifies category two validation surrounding: Family Affiliation, Citizenship and Closure reason. Further discussion with OFA indicated that it is important/a priority for a STT to see all category two errors encompassing these fields in their entirety. That makes prioritizing these category two errors extremely easy because the need to query those fields by specific values and expected values is not required. The queries below provide a complete implementation to query all category two errors encompassing those fields.
+
+```python
+# All cat2 errors associated with FAMILY_AFFILIATION and (CITIZENSHIP_STATUS or CLOSURE_REASON)
+second_field = "CITIZENSHIP_STATUS" if is_active else "CLOSURE_REASON"
+field_query = Q(field_name="FAMILY_AFFILIATION") | Q(field_name=second_field)
+filtered_errors = filtered_errors.union(all_errors.filter(
+    field_query,
+    error_type=ParserErrorCategoryChoices.FIELD_VALUE
+    ))
+```
+
+The key insight for the category three case is less obvious. Looking at the OneNote, it seems as though we might need to query errors based on field name(s), expected value and actual value. However, for category three errors that information is encoded into the error by its existence. For example, the OneNote indicates that a high priority error a STT should have included in their report is `If fam affil = 1 then SSN must be valid `. This exact error and it's values (expected and given) can be uniquely found in any of the active or closed case record schemas. E.g.:
```

Comment:
pretty accurate, from a technical perspective. For non-technical folks, the critical errors are related to how OFA defines cash assistance receipt, and work-eligible individuals.